Lib/functools.py

```diff
@@ -16,7 +16,34 @@
 try:
     from _functools import reduce
 except ImportError:
-    pass
+    # fallback implementation in case C module is unavailable
+    _initial_missing = object()
+    def reduce(function, iterable, initializer=_initial_missing):
+        """reduce(function, sequence[, initial]) -> value
+
+            Apply a function of two arguments cumulatively to the items of a sequence,
+            from left to right, so as to reduce the sequence to a single value.
+            For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
+            ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
+            of the sequence in the calculation, and serves as a default when the
+            sequence is empty.
+        """
+        try:
+            iterable = iter(iterable)
+        except TypeError as err:
+            raise TypeError('reduce() arg 2 must support iteration')
```

Comment:
```suggestion
            raise TypeError('reduce() arg 2 must support iteration') from None
```
You don't want the first TypeError to show up in the traceback